runner/service_instance.go

```diff
@@ -197,24 +200,95 @@ func (s *ServiceInstance) Error() error {
 	return s.runErr
 }
 
-func (s *ServiceInstance) Stop(sig syscall.Signal) error {
+func (s *ServiceInstance) Stop() error {
+	var signal syscall.Signal
+	switch s.ShutdownSignal {
+	case "SIGKILL":
+		signal = syscall.SIGKILL
+	case "SIGTERM":
+		signal = syscall.SIGTERM
+	default:
+		// Default to SIGKILL if unspecified or unrecognized. In case we add new values to itest.bzl but forget to add it here
+		signal = syscall.SIGKILL
+	}
+
+	return s.StopWithSignal(signal)
+}
+
+func isGone(err error) bool {
+	if errors.Is(err, os.ErrProcessDone) {
+		return true
+	}
+
+	var errno syscall.Errno
+	return errors.As(err, &errno) && errno == syscall.ESRCH
+}
+
+func (s *ServiceInstance) StopWithSignal(signal syscall.Signal) error {
+	if s.cmd.Process == nil {
+		return nil
+	}
+
+	err := killGroup(s.cmd, signal)
+	if isGone(err) {
+		return nil
+	}
+
 	func() {
 		s.mu.Lock()
 		defer s.mu.Unlock()
 		s.killed = true
 	}()
 
-	if s.cmd.Process == nil {
-		return nil
-	}
-
-	err := killGroup(s.cmd, sig)
 	if err != nil {
 		return err
 	}
 
-	for !s.isDone() {
-		time.Sleep(5 * time.Millisecond)
+	if signal == syscall.SIGKILL {
+		log.Printf("Sent SIGKILL to %s\n", s.Colorize(s.Label))
+		for !s.isDone() {
+			time.Sleep(5 * time.Millisecond)
+		}
+	} else {
+		log.Printf("Sent SIGTERM to %s\n", s.Colorize(s.Label))
+
+		shutdownTimeout, err := time.ParseDuration(s.VersionedServiceSpec.ShutdownTimeout)
+		if err != nil {
+			log.Printf("failed to parse health check timeout, falling back to no timeout: %v", err)
+			return err
+		}
+
+		log.Printf("Sent SIGTERM to %s, waiting for %s to stop gracefully\n", s.Colorize(s.Label), s.VersionedServiceSpec.ShutdownTimeout)
+
+		waitFor := time.After(shutdownTimeout)
+		for {
+			// Check if the process has exited
+			if s.isDone() {
+				break
+			}
+
+			select {
+			case <-waitFor:
+				log.Printf("%s did not exit within %s, sending SIGKILL. If you are trying to collect coverage, you will most likely miss stats, try increasing the shutdown timeout.\n", s.Colorize(s.Label), shutdownTimeout)
+
+				err := killGroup(s.cmd, syscall.SIGKILL)
+				if err != nil {
```

Comment:
technically this should also check `isGone` in case the process finally exited correctly while we were delivering the signal